openapi3/internalize_refs.go

```diff
@@ -0,0 +1,389 @@
+package openapi3
+
+import (
+	"context"
+	"path/filepath"
+	"strconv"
+	"strings"
+)
+
+type RefNameResolver func(string) string
+
+// DefaultRefResolver is a default implementation of refNameResolver for the
+// InternalizeRefs function.
+//
+// If a reference points to an element inside a document, it returns the last
+// element in the reference using filepath.Base. Otherwise if the reference points
+// to a file, it returns the file name trimmed of all extensions.
+func DefaultRefNameResolver(ref string) string {
+	if ref == "" {
+		return ""
+	}
+	split := strings.SplitN(ref, "#", 2)
+	if len(split) == 2 {
+		return filepath.Base(split[1])
+	}
+	ref = split[0]
+	for ext := filepath.Ext(ref); len(ext) > 0; ext = filepath.Ext(ref) {
+		ref = strings.TrimSuffix(ref, ext)
+	}
+	return filepath.Base(ref)
+}
+
+func schemaNames(s Schemas) []string {
+	out := make([]string, 0, len(s))
+	for i := range s {
+		out = append(out, i)
+	}
+	return out
+}
+
+func parametersMapNames(s ParametersMap) []string {
+	out := make([]string, 0, len(s))
+	for i := range s {
+		out = append(out, i)
+	}
+	return out
+}
+
+func isExternalRef(ref string) bool {
+	return ref != "" && !strings.HasPrefix(ref, "#/components/")
+}
+
+func (doc *T) addSchemaToSpec(s *SchemaRef, refNameResolver RefNameResolver) string {
+	if s == nil || !isExternalRef(s.Ref) {
+		return ""
+	}
+
+	name := refNameResolver(s.Ref)
+
+	// I just added a more comprehensive check for this being a schema but people should really
+	// make sure their resolver returns unique names
+	val := 1
+	for {
+		// basic check, may need something more exhaustive...
+		if existing, ok := doc.Components.Schemas[name]; ok {
+			if existing.Value.Description != s.Value.Description || s.Value.Type != existing.Value.Type {
+				name = refNameResolver(s.Ref) + strconv.Itoa(val)
+				val++
+				continue
+			}
+			newRef := "#/components/schemas/" + name
+			if newRef == s.Ref {
+				s.Ref = ""
+			} else {
+				s.Ref = newRef
+			}
+			return s.Ref
+		}
+		break
+	}
+	if doc.Components.Schemas == nil {
+		doc.Components.Schemas = make(Schemas)
+	}
+	doc.Components.Schemas[name] = s.Value.NewRef()
+	s.Ref = "#/components/schemas/" + name
+	return s.Ref
+}
+
+func (doc *T) addParameterToSpec(p *ParameterRef, refNameResolver RefNameResolver) string {
+	if p == nil || !isExternalRef(p.Ref) {
+		return ""
+	}
+	name := refNameResolver(p.Ref)
+	if _, ok := doc.Components.Parameters[name]; ok {
+		p.Ref = "#/components/parameters/" + name
+		return p.Ref
+	}
+
+	if doc.Components.Parameters == nil {
+		doc.Components.Parameters = make(ParametersMap)
+	}
+	doc.Components.Parameters[name] = &ParameterRef{Value: p.Value}
+	p.Ref = "#/components/parameters/" + name
+	return p.Ref
+}
+
+func (doc *T) addHeaderToSpec(h *HeaderRef, refNameResolver RefNameResolver) string {
```

Comment:
```suggestion
func (doc *T) addHeaderToSpec(h *HeaderRef, refNameResolver RefNameResolver) {
```
No call to this function actually use the returned value so let's drop that.